docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,244 @@
+
+> 4 . After 15 minutes (or time configured by `syncDelay`), if there is no fully loadable `meta.json`, we assume block is malformed, partially uploaded or marked for deletion and can be
+*removed* by compactor. [Compactor change needed]
+
+This is to detect partial uploads caused by for example compactor being rolled out or crashed. Other writers managed by `shipper` package relies on 
+fact that blocks are persisted on disk so the upload operation can be retried.
+
+Similar to writers, readers are excluding all blocks fresher than 15m or older but without `meta.json`. Overall those 4 rules
+solved issue C) and A), B) in terms of block upload. 
+
+However, sometimes we need to change blocks for reasons, like:
+
+* compaction, so merging couple of blocks into one with shared index.
+* potentially in future: `delete series` operations.
+* repair operations (e.g fixes against Prometheus or Thanos bugs like index issues).
+
+Since all blocks are immutable, such changes are performed by rewriting block (or compacting multiple) into new block. This 
+makes `removal` operations a first citizen object lifecycle. The main difficult point during this process is to make sure all 
+readers are synced and aware that new block is ready in place of the old one(s), so writer can remove old block(s). In eventual consistency system
+we don't have that information without additional coordination. To mitigate we propose 2 new rules. All to support lazy deletion:
+
+> 5 . Overlaps (in external labels and time dimensions) with the same data are allowed and should be handled by readers gracefully. [Compactor change needed]
+
+This should be the case currently for all components except compactor. For example store having overlapping blocks with even the same data
+or sidecar and store exposing same data via StoreAPI is handled gracefully by frontend.
+
+The major difference here is compactor. Currently compactor does not allow overlap. It immediately halts compactor and waits for 
+manual actions. This is on purpose to not allow block malformation by blocks which should be not compacted together.
+
+"The same data" overlap is crucial here. The overlap with same data (valid overlap) is when:
+
+*Newer block from two or more overlapping blocks fully submatches the source blocks of older blocks. Older blocks can be then ignored.*
+
+Fully word is crucial. For example we won't be able to resolve case with block ABCD and CDEF. This is because there is no logic for decompact or vertical compaction.
+
+Having this kind of overlap support, we can delay deletion by forming 6th rule:
+
+> 6 . Adding "replacement" or "overlapped" block causes old blocks to be scheduled for removal when new block is ready by compactor (`HealthyOverlapped` state). [Compactor/Repair change needed]
+
+This rule ensures that we can detect when to delete block, and we delete it only once new block is ready, so when we mitigated eventual consistency: all readers should be aware of new block.
```

Comment:
`so when we mitigated eventual consistency` , do you mean , eg: `this will mitigate eventual consistency`?